Fix various leaks and reference cycles in Web Extension classes.

xeenon opened the pull request:
#### 01ba8487b4464ec9fe00767613438148240e234f
<pre>
Fix various leaks and reference cycles in Web Extension classes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248207">https://bugs.webkit.org/show_bug.cgi?id=248207</a>

Reviewed by Darin Adler.

Issues this fixes:
* WKWebViewConfiguration holds a references to _WKWebExtensionController, and that strong references
  was causing a retain cycle when WebExtensionContext made a configuration and loaded a WKWebView.
  To solve this, add a new _weakWebExtensionController property to the configuration and page proxy
  that can be used internally when making extension WKWebViews.
* WebExtensionController was not unloading contexts on destruct, and that was causing some world leaks.
* WebProcess proxy objects were not removing themselves from the MessageReceiverMap, causing an ASSERT.
  This wasn&apos;t fatal, since the MessageReceiverMap uses WeakPtr now and the count is just a debug tool.
  Maybe we should remove the MessageReceiverMap counting now that WeakPtr makes it moot?
* Removed empty NSSecureCoding support from _WKWebExtensionController since it did nothing and really
  would need a complex implementation make it work correctly. This was conflicting with the WeakPtr changes.

* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::copy const): Added weakWebExtensionController.
(API::PageConfiguration::weakWebExtensionController): Added.
(API::PageConfiguration::setWeakWebExtensionController): Added.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]): Setup weakWebExtensionController.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration encodeWithCoder:]): Removed webExtensionController.
(-[WKWebViewConfiguration initWithCoder:]): Ditto.
(-[WKWebViewConfiguration copyWithZone:]): Ditto.
(-[WKWebViewConfiguration _strongWebExtensionController]): Added.
(-[WKWebViewConfiguration _webExtensionController]): Support returning _weakWebExtensionController.
(-[WKWebViewConfiguration _weakWebExtensionController]): Added.
(-[WKWebViewConfiguration _setWeakWebExtensionController:]): Added.
(-[WKWebViewConfiguration _webExtensionControllerIfExists]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController loadExtensionContext:]): Drive-by add of NSParameterAssert to match other methods.
(-[_WKWebExtensionController loadExtensionContext:error:]): Ditto.
(-[_WKWebExtensionController unloadExtensionContext:]): Ditto.
(-[_WKWebExtensionController unloadExtensionContext:error:]): Ditto.
(-[_WKWebExtensionController extensionContextForExtension:]): Ditto.
(+[_WKWebExtensionController supportsSecureCoding]): Deleted.
(-[_WKWebExtensionController initWithCoder:]): Deleted.
(-[_WKWebExtensionController encodeWithCoder:]): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::webViewConfiguration): Set _weakWebExtensionController property.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::unload): Change unload order and protect context.
(WebKit::WebExtensionController::unloadAll): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask): Early return if weak WebExtensionController is null.
(WebKit::WebExtensionURLSchemeHandler::platformTaskCompleted): Added. Tasks were not being removed from map.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::~WebExtensionController): Added. Call unloadAll().
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionURLSchemeHandler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy): Added m_weakWebExtensionController.
(WebKit::WebPageProxy::close): Use m_weakWebExtensionController too.
(WebKit::WebPageProxy::creationParameters): Ditto.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::~WebExtensionContextProxy): Added. Call removeMessageReceiver().
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.cpp:
(WebKit::WebExtensionControllerProxy::~WebExtensionControllerProxy): Added. Call removeMessageReceiver().
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:

Canonical link: <a href="https://commits.webkit.org/256956@main">https://commits.webkit.org/256956@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4de372021708137360995782a248349c3da8464

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97349 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6612 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30509 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106867 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6918 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35351 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/89758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103545 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103013 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83979 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/89758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/89758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/628 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/610 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5414 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2357 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1855 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/41141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -92,6 +92,8 @@
     if (outError)
         *outError = nil;
 
+    Ref<WebExtensionContext> protectedExtensionContext = extensionContext;
```
Should be able to use just `Ref` here without a template argument because of deduction guides.

darinadler commented on Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
```diff
@@ -55,6 +55,13 @@ WebExtensionController::WebExtensionController()
 
+WebExtensionController::~WebExtensionController()
+{
+#if PLATFORM(COCOA)
```
Probably needs a comment about why this is Cocoa-only.

darinadler reviewed (approved):
Seems a little inelegant to need the pair of data members but I do think this is an effective approach.

xeenon committed:
Fix various leaks and reference cycles in Web Extension classes.
https://bugs.webkit.org/show_bug.cgi?id=248207

Reviewed by Darin Adler.

Issues this fixes:
* WKWebViewConfiguration holds a references to _WKWebExtensionController, and that strong references
  was causing a retain cycle when WebExtensionContext made a configuration and loaded a WKWebView.
  To solve this, add a new _weakWebExtensionController property to the configuration and page proxy
  that can be used internally when making extension WKWebViews.
* WebExtensionController was not unloading contexts on destruct, and that was causing some world leaks.
* WebProcess proxy objects were not removing themselves from the MessageReceiverMap, causing an ASSERT.
  This wasn't fatal, since the MessageReceiverMap uses WeakPtr now and the count is just a debug tool.
  Maybe we should remove the MessageReceiverMap counting now that WeakPtr makes it moot?
* Removed empty NSSecureCoding support from _WKWebExtensionController since it did nothing and really
  would need a complex implementation make it work correctly. This was conflicting with the WeakPtr changes.

* Source/WebKit/UIProcess/API/APIPageConfiguration.cpp:
(API::PageConfiguration::copy const): Added weakWebExtensionController.
(API::PageConfiguration::weakWebExtensionController): Added.
(API::PageConfiguration::setWeakWebExtensionController): Added.
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _setupPageConfiguration:]): Setup weakWebExtensionController.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration encodeWithCoder:]): Removed webExtensionController.
(-[WKWebViewConfiguration initWithCoder:]): Ditto.
(-[WKWebViewConfiguration copyWithZone:]): Ditto.
(-[WKWebViewConfiguration _strongWebExtensionController]): Added.
(-[WKWebViewConfiguration _webExtensionController]): Support returning _weakWebExtensionController.
(-[WKWebViewConfiguration _weakWebExtensionController]): Added.
(-[WKWebViewConfiguration _setWeakWebExtensionController:]): Added.
(-[WKWebViewConfiguration _webExtensionControllerIfExists]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController loadExtensionContext:]): Drive-by add of NSParameterAssert to match other methods.
(-[_WKWebExtensionController loadExtensionContext:error:]): Ditto.
(-[_WKWebExtensionController unloadExtensionContext:]): Ditto.
(-[_WKWebExtensionController unloadExtensionContext:error:]): Ditto.
(-[_WKWebExtensionController extensionContextForExtension:]): Ditto.
(+[_WKWebExtensionController supportsSecureCoding]): Deleted.
(-[_WKWebExtensionController initWithCoder:]): Deleted.
(-[_WKWebExtensionController encodeWithCoder:]): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::webViewConfiguration): Set _weakWebExtensionController property.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::unload): Change unload order and protect context.
(WebKit::WebExtensionController::unloadAll): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask): Early return if weak WebExtensionController is null.
(WebKit::WebExtensionURLSchemeHandler::platformTaskCompleted): Added. Tasks were not being removed from map.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.cpp:
(WebKit::WebExtensionController::~WebExtensionController): Added. Call unloadAll().
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionURLSchemeHandler.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy): Added m_weakWebExtensionController.
(WebKit::WebPageProxy::close): Use m_weakWebExtensionController too.
(WebKit::WebPageProxy::creationParameters): Ditto.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::~WebExtensionContextProxy): Added. Call removeMessageReceiver().
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.cpp:
(WebKit::WebExtensionControllerProxy::~WebExtensionControllerProxy): Added. Call removeMessageReceiver().
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:

Canonical link: https://commits.webkit.org/256956@main
